### PR TITLE
Lighter memoize with OneShot

### PR DESCRIPTION
**Pull request by regiskuckaertz**

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -710,13 +710,17 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
...
-      _ <- (l.await *> ((self provide r) to p)).fork
-    } yield l.succeed(()) *> p.await
+  final def memoize: URIO[R, IO[E, A]] = {
+    import internal.OneShot
+    (ZIO.environment[R] <*> UIO.effectTotal(OneShot.make[A])).map { case (r, x) =>
+      IO.effectTotal {
+        if (x.isSet)
+          UIO.effectTotal(x.get)
+        else
+          self.tap(a => UIO.effectTotal(x.set(a))).provide(r)
```
Comment (iravid) [inline core/shared/src/main/scala/zio/ZIO.scala]: `OneShot#set` throws if called after the value has been set, but we don't have anything guaranteeing it won't be called twice here, if I'm not mistaken.

Comment (regiskuckaertz) [reply core/shared/src/main/scala/zio/ZIO.scala]: You mean in sth like `action.memoize.flatMap { a => a <&> a }`? Yeah 🤦‍♂

Comment (iravid) [reply core/shared/src/main/scala/zio/ZIO.scala]: Maybe with a `Ref[Either[A, Promise]]` state machine?

**Comment by regiskuckaertz** [comment]
Thought I'd find something more elegant but admittedly failed.

**Comment by regiskuckaertz** [comment]
@iravid Yeah, I switched back to `Ref` but had to create a promise each time, which is ugly. Thanks a lot for reviewing 🤗

**Pull request closed without merging**